Embedded schemas

klaussfreire opened the pull request:
Allow schemas to be picklable, and embed a pickled schema inside
MappedArrays, to make mapping safe across client code versions

klaussfreire committed:
Embedded schemas

Allow schemas to be picklable, and embed a pickled schema inside
MappedArrays, to make mapping safe across client code versions

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1112,20 +1112,43 @@ def Proxy(self):
+    def __setstate__(self, state):
+        # These are a safety hazard (they could cause OOM if corrupted), so ignore them if present
+        # Not needed for unpacking anyway
+        state.pop('pack_buffer_size', None)
```
I am not sure about this... If during the `__init__.py` it uses a different value than the default, then that value is going to be lost because the `init` uses the default method.

klaussfreire replied:
Well, that's the whole point. Pack buffer size only matters for packing, we want to pickle schemas for unpacking purposes only, and also pack buffer size controls memory allocation directly - we don't want external input controlling memory allocation like that.

tzulberti-jampp commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1547,6 +1588,22 @@ def __init__(self, buf, offset = 0, idmap = None, idmap_size = 1024):
+            elif self.schema is None:
+                raise ValueError("Cannot map schema-less buffer without specifying schema")
+        else:
+            self.version = 0
```
I couldn't find were the `version` attribute is used.

klaussfreire replied:
It's not. I just thought it might be interesting some day to know.

klaussfreire replied:
You know... on-disk format alterations aren't simple, so better add now what we think might be useful later.

zeitos replied:
you mean know it from previous versions? ... how would that work here? b/c that makes sense for logging extra values but I'm not sure there's much value in keeping unused code here.

zeitos replied:
lets talk this in person.

klaussfreire replied:
I was talking about the version field in the on-disk format.

The attribute also makes sense, to have it for future code if it needs to know the version somehow. Note that future code could be external code. So it's an interface added because it seems useful, even if we make no immediate use of it.

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1112,20 +1112,43 @@ def Proxy(self):
+
+        for k in self.slot_keys:
+            if self.slot_types[k] is not other_schema.slot_types[k]:
+                return False
```
Consider using all
```
all(k in other_schema.slot_types and other_schema.slot_types[k] == v for k, v in self.slotkeys.items())
```

klaussfreire replied:
Using all there is complex. All takes a generator/list, and that slot_types there is a private attribute of the Schema only accessible through Cython. I'm pretty sure to make it work it would either be ugly (making a type-annotated generator that would take 4x the line count) or inefficient (building a list)

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1607,14 +1664,16 @@ def __len__(self):
         initial_pos = destfile.tell()
         write = destfile.write
         write(cls._Header.pack(0, 0, 0))
+        write(cls._NewHeader.pack(cls._CURRENT_VERSION, 0, 0))
```
What happens if it's an older or newer version?

klaussfreire replied:
We don't write any other versions than the current

zeitos replied:
but what happen when you call pack with an older version?

zeitos commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1642,9 +1701,15 @@ def build(cls, initializer, destfile = None, tempdir = None, idmap = None):
         final_pos = destfile.tell()
         destfile.seek(initial_pos)
         write(cls._Header.pack(final_pos - initial_pos, index_pos - initial_pos, len(index)))
+        write(cls._NewHeader.pack(cls._CURRENT_VERSION, schema_pos - initial_pos, final_pos - schema_pos))
```
same here...

klaussfreire replied:
Same again, while writing, we only care about the current version

klaussfreire committed:
Merge branch 'master' of github.com:jampp/sharedbuffers into embedded_schemas

klaussfreire committed:
Whitespace fix

klaussfreire committed:
Add minimum reader version to binary format

To ensure old readers don't attempt to open a future new format,
if it ends up being incompatible

klaussfreire committed:
Reject version-0 buffers

Attribute sort order changed relative to version-0, and version-0
has no embedded schema, so we can't really open them
